mava/networks/utils/sable/encoder_decoder_fns.py

```diff
@@ -0,0 +1,271 @@
+    legal_actions: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, Tuple[chex.Array, chex.Array]]:
+    logit = jnp.zeros_like(legal_actions, dtype=jnp.float32)
+
+    # Apply the decoder per chunk
+    num_chunks = shifted_actions.shape[1] // chunk_size
+    for chunk_id in range(0, num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = (chunk_id + 1) * chunk_size
+        chunked_obs_rep = obs_rep[:, start_idx:end_idx]
+        chunk_shifted_actions = shifted_actions[:, start_idx:end_idx]
+        chunk_dones = dones[:, start_idx:end_idx]
+        chunk_step_count = step_count[:, start_idx:end_idx]
+        chunk_logit, hstates = decoder(
+            action=chunk_shifted_actions,
+            obs_rep=chunked_obs_rep,
+            hstates=hstates,
+            dones=chunk_dones,
+            step_count=chunk_step_count,
+        )
+        logit = logit.at[:, start_idx:end_idx].set(chunk_logit)
+
+    return logit, hstates
+
+
+def get_shifted_actions(action: chex.Array, legal_actions: chex.Array, n_agents: int) -> chex.Array:
+    """Get the shifted action sequence for predicting the next action."""
+    # Get the batch size, sequence length, and action dimension
+    B, S, N = legal_actions.shape
+
+    # Create a shifted action sequence for predicting the next action
+    # Initialize the shifted action sequence.
+    shifted_actions = jnp.zeros((B, S, N + 1))
+
+    # Set the start-of-timestep token (first action as a "start" signal)
+    start_timestep_token = jnp.zeros(N + 1).at[0].set(1)
+
+    # One hot encode the action
+    one_hot_action = jax.nn.one_hot(action, N)
+
+    # Insert one-hot encoded actions into shifted array, shifting by 1 position
+    shifted_actions = shifted_actions.at[:, :, 1:].set(one_hot_action)
+    shifted_actions = jnp.roll(shifted_actions, shift=1, axis=1)
+
+    # Set the start token for the first agent in each timestep
+    shifted_actions = shifted_actions.at[:, ::n_agents, :].set(start_timestep_token)
+
+    return shifted_actions
+
+
+def init_sable(
+    encoder: nn.Module,
+    decoder: nn.Module,
+    obs_carry: chex.Array,
+    hstates: chex.Array,
+    key: chex.PRNGKey,
+) -> chex.Array:
+    """Initializating the network: Applying non chunkwise encoding-decoding."""
+    # Get the observation, legal actions, and timestep id
+    obs, legal_actions, step_count = (
+        obs_carry.agents_view,
+        obs_carry.action_mask,
+        obs_carry.step_count,
+    )
+
+    # Apply the encoder
+    v_loc, obs_rep, _ = execute_encoder_fn(
+        encoder=encoder,
+        obs=obs,
+        decayed_hstate=hstates[0],
+        step_count=step_count,
+        chunk_size=obs.shape[1],
+    )
+
+    # Apply the decoder
+    _ = autoregressive_act(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        legal_actions=legal_actions,
+        hstates=hstates[1],
+        step_count=step_count,
+        key=key,
+    )
+
+    return v_loc
+
+
+def execute_encoder_fn(
```

Comment:
Can we use act instead of execute
```suggestion
def act_encoder_fn(
```